_posts/2025-01-03-Clang-17-now-in-dev-container.markdown

```diff
@@ -0,0 +1,56 @@
+---
+layout: post
+title:  "Clang 17 now in dev container, and other toolchain news"
+date:   2025-01-23
+categories: cheri toolchain
+author: Owen Anderson
+---
+
+Thanks to a collaboration with the [upstream CHERI toolchain](https://github.com/ctsrd-cheri/llvm-project), the CHERIoT toolchain has now been rebased onto Clang 17 from Clang 13, bringing us two years closer to upstream Clang.
+Major thanks to [Alex Richardson](https://github.com/arichardson) and [Sam Leffler](https://github.com/sleffler/) for their work on this effort!
+
+This updates brings with it substantially improved support for C++20 features, as well as preliminary support for some C++23 and C23 features.
+It also brings with it many improvements to the core of the RISC-V code generator, notably also benefiting code size: the firmware image for the Cheriot RTOS testsuite is 2.7% smaller when built with Clang 17 compared to Clang 13.
+Other highlights include compile time improvements, and too many under-the-hood fixes to enumerate.
+You can find detailed release notes for all Clang and LLVM releases on the LLVM.org [releases page](https://releases.llvm.org).
+
+## Other Toolchain Improvements
+
+Since landing the Clang 17 rebase, we've been busy bringing bugfixes and enhancements to the CHERIoT toolchain, including:
+
+### Language & Usability Improvements
+
+- [Implemented](https://github.com/CHERIoT-Platform/llvm-project/commit/025c5d452e8935ebbe2a09d78fb2a10c1c96a626) a new Clang diagnostic to warn on compartment exports that return void, or where the return value is unused.
+These warnings are off by default until Cheriot RTOS has been updated for them, and are controlled by the `-Wcheri-compartment-return-void` compiler flag. Thanks to [Robert Dazi](https://github.com/v01dXYZ) for this one!
+- [Allowed](https://github.com/CHERIoT-Platform/llvm-project/commit/0de0fb3e8f63be9102c5b5eab1b496415b667ca9) `cheri_libcall`-annotated functions to decay into unannotated function pointers.
+This is useful for passing the address of a `cheri_libcall` function as a callback.
+- [Improved](https://github.com/CHERIoT-Platform/llvm-project/commit/b14e86345d929bf91ab3fb1197ac716dc7ca6e2d) linker error reporting if you accidentally omit the compartment export annotation on a declaration.
+LLD will now look for matching unexported functions and provide a suggested fix.
```

Comment:
I tried to trigger this one in the book and couldn't.  Can you take a look at text/examples/compartment_annotation/ in the book and see why removing the include of interface.h from one of the compartments that defines an export doesn't trigger this helpful error?

Reply:
I wrote a synthetic example when developing it that I confirmed worked. I can take a look at why it's not triggering here.

Reply:
I am able to trigger the lld warning with this code change: https://github.com/CHERIoT-Platform/book/pull/10

```
ld.lld: error: undefined symbol: monotonic(int (int) chericcall*)
>>> referenced by stdio.h:56 (/Users/resistor/Documents/cheriot-book/rtos-source/sdk/include/stdio.h:56)
>>>               build/cheriot/cheriot/release/entry.compartment:()
>>> referenced by stdio.h:57 (/Users/resistor/Documents/cheriot-book/rtos-source/sdk/include/stdio.h:57)
>>>               build/cheriot/cheriot/release/entry.compartment:()
>>> did you mean the "monotonic(int (int) chericcall*)" export from compartment "example_compartment"?
>>> defined in: build/cheriot/cheriot/release/example_compartment.compartment
>>> the declaration may be missing a compartment annotation
```